spectator-api/src/main/java/com/netflix/spectator/api/ArrayTagSet.java

```diff
@@ -184,11 +185,12 @@ ArrayTagSet addAll(Map<String, String> ts) {
       String[] newTags = new String[2 * ts.size()];
       int i = 0;
       for (Map.Entry<String, String> entry : ts.entrySet()) {
-        newTags[i++] = entry.getKey();
-        newTags[i++] = entry.getValue();
+        newTags[i++] = Objects.requireNonNull(entry.getKey(), "tag keys cannot be null");
+        newTags[i++] = Objects.requireNonNull(entry.getValue(), "tag values cannot be null");
       }
-      checkForNullValues(newTags);
-      return addAll(newTags, newTags.length);
+
+      boolean sorted = ts instanceof SortedMap && ((SortedMap<?, ?>) ts).comparator() == null;
```

Comment:
Should this also check if the natural ordering is explicit? 

```java
private boolean hasNaturalOrder(SortedMap<String, String> ts) {
  return ts.comparator() == null || ts.comparator() == Comparator.<String>naturalOrder();
}
```

That would cover usage like `new TreeMap<>(Comparator.naturalOrder())`, though not `new TreeMap<>(String::compareTo)`.

Reply:
I thought about checking for `Comparator.naturalOrder()`, although I think it might be unusual to explicitly supply that comparator since the Javadoc for SortedMap / TreeMap indicate natural ordering is the default. I can go ahead and add a check for that anyway, but I think it's not possible to cover all scenarios (Guava collections use an explicit comparator for natural ordering that is different than `Comparator.naturalOrder()`, and `String::compareTo` like you mentioned.

Reply:
Okay, I've updated the PR to check for `Comparator.naturalOrder()` as well.